src/upgrades/mod.rs

```diff
@@ -159,13 +166,109 @@ pub trait UpgradeStore {
+            "Scanning for not yet mapped OpenSSL signer keys in {} to record in the signer store",
+            keys_dir.to_string_lossy()
+        );
+
+        // For every file (key) in the legacy OpenSSL signer keys directory
+        if let Ok(dir_iter) = keys_dir.read_dir() {
+            let mut krill_signer: Option<KrillSigner> = None;
+            let mut signer_handle: Option<Handle> = None;
+
+            for entry in dir_iter {
+                let entry = entry.map_err(|err| {
+                    UpgradeError::IoError(KrillIoError::new(
+                        format!(
+                            "I/O error while looking for signer keys to register in: {}",
+                            keys_dir.to_string_lossy()
+                        ),
+                        err,
+                    ))
+                })?;
+
+                if entry.path().is_file() {
+                    // Is it a key identifier?
+                    if let Ok(key_id) = KeyIdentifier::from_str(&entry.file_name().to_string_lossy()) {
+                        if krill_signer.is_none() {
+                            // Create the KrillSigner which will in turn create the actual signers defined in the
+                            // users Krill configuration file. If the keys in the legacy OpenSSL signer keys directory
+                            // are actively being used by Krill the config must cause KrillSigner to instantiate the
+                            // corresponding OpenSSL signer that owns those keys.
+                            krill_signer = Some(KrillSigner::build(&config.data_dir, config.signers())?);
+                        }
+
+                        if let Some(krill_signer) = &krill_signer {
+                            // Is the key already recorded in the mapper? It shouldn't be, but asking will cause the
+                            // initial registration of the OpenSSL signer to occur and for it to be assigned a handle.
+                            // We need the handle so that we can register keys with the mapper.
+                            if !krill_signer.get_key_info(&key_id).is_ok() {
+                                // No, record it
+
+                                // Do we already know the handle of the OpenSSL signer? If not, find it. It should be
+                                // possible to find the key via the OpenSSL signer because it shoud look for it in the
+                                // same directory and file name as we are now processing. Finding it confirms for us
+                                // that we have the correct signer handle (unless someone migrated the keys into another
+                                // signer and registered them there and then manually deleted the signer store directory
+                                // ... but then if they go behind Krill's back there are limits to what we can cope with)
+                                if signer_handle.is_none() {
```

Comment:
In the current HSM code it is possible to use Krill without any OpenSSL signer at all. By default if you configure a PKCS#11 or KMIP signer then an OpenSSL signer will be used for one-off key generation & signing, and for random number generation but it's possible to configure Krill to use the backend HSM for both of these cases instead of OpenSSL. Aactually I need to look at the random number generation case as there is a subtle difference between using OpenSSL always to generate random numbers and using it only in case the backend HSM doesn't support random number generation...

Reply:
Hmm, I guess you are right actually. Krill doens't need an OpenSSL signer for fresh installs, but this code runs in the context of a pre-existing instance of Krill that has at least one OpenSSL key already, therefore it MUST have had an OpenSSL signer to create those keys.

Reply:
One caveat however: it's possible that as part of upgrading a user edited their `krill.conf` so that it defines two OpenSSL signers each with their own keys directory, e.g. for testing keyroll between signers. We then can't "find" the signer by assuming it is the one and only OpenSSL signer. That's why this code is looking for the signer that owns the key rather than for the first OpenSSL signer.